lib/src/test/java/com/knuddels/jtokkit/reference/Cl100kBaseTest.java

```diff
@@ -0,0 +1,234 @@
+import java.util.stream.IntStream;
+
+import static java.lang.Character.*;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.stream.Collectors.joining;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class Cl100kBaseTest {
+    public static final String PUNCTUATION = "'\".,?!:()";
+    private static final String LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZő你好ſ ½";
+    private static final String NUMBERS = "0123456789½";
+    private static final String NEWLINES = "\n\r";
+    private static final String WHITESPACES = " \t " + NEWLINES;
+    private static final String NOT_NEWLINE_OR_LETTER_OR_NUMERIC = " \t🤚🏾😩" + PUNCTUATION;
+    private static final String NOT_WHITESPACE_OR_LETTER_OR_NUMERIC = NOT_NEWLINE_OR_LETTER_OR_NUMERIC + NEWLINES;
+    private static final List<String> SPECIAL = List.of("'s", "'t", "'re", "'ve", "'m", "'ll", "'d", "'ſ", "'x", "🤚🏾", "😩", "　", "½");
+
+    private static final Encoding ENCODING = EncodingFactory.cl100kBase();
+
+    static String normalizeStringForTesting(String testString) {
+        return testString
+                .replaceAll("\\r", "\\\\r")
+                .replaceAll("\\n", "\\\\n")
+                .replaceAll(" ", "␣");
+    }
+
+    private static ThreadLocalRandom rand() {
+        return ThreadLocalRandom.current();
+    }
+
+    private static String getMessage(String testString, Encoding originalEncoder, Encoding encoder, int maxTokenCount) {
+        var expected = originalEncoder.encode(testString, maxTokenCount).getTokens();
+        var actual = encoder.encode(testString, maxTokenCount).getTokens();
+        return "`" + testString + "` should have mapped to: " + expected + " for maxTokenCount = " + maxTokenCount + ", but was: " + actual;
+    }
+
+    @Disabled
+    @Test
+    public void measureEncodingSpeeds() {
+        var input = new StringBuilder();
+        var measurements = new TreeMap<Integer, Long>();
+
+        var iterations = 20;
+        for (var i = 1.0; i < 3_000; i = Math.max(i + 1, i * 1.01)) {
+            while (input.length() < i) {
+                input.append("a");
+            }
+            var inputString = input.toString();
+
+            for (var j = 0; j < 10 * iterations; j++) {
+                var warmup = ENCODING.encode(inputString);
+                assert !warmup.isEmpty();
+            }
+            var startTime = System.nanoTime();
+            for (var j = 0; j < iterations; j++) {
+                var encodingResult = ENCODING.encode(inputString);
+                assert !encodingResult.isEmpty();
+            }
+            var endTime = System.nanoTime();
+            measurements.put((int) i, ((endTime - startTime) / iterations));
+        }
+        measurements.forEach((i, t) -> System.out.println(i + "\t" + t));
+    }
+
+    @Test
+    void testEdgeCaseRoundTrips() throws Exception {
```

Comment:
a few problematic cases that weren't covered by other test-cases that I've encountered during the optimization process